File: LeetCode/Medium/Ugly Number II/Go/ugly_number_two.go
package main

import "fmt"

//----------brute force---------------
func is_ugly(n int) bool {
	if n == 0 {
		return false
	}
	for n%2 == 0 {
		n = n / 2
	}
	for n%3 == 0 {
		n = n / 3
	}
	for n%5 == 0 {
		n = n / 5
	}
	return n == 1
}

func ugly_number_two(n int) int {
	cnt := 1
	i := 1
	for cnt <= n {
		if is_ugly(i) {
			cnt++
		}
		i++
	}
	return i - 1
}

//----------brute force---------------

//----------efficient approach---------------
func ugly_number_two_eff(n int) int {
	dp := make([]int, n)
	dp[0] = 1
	count := 1
	idx_2, idx_3, idx_5 := 0, 0, 0
	for count < n {
		v1, v2, v3 := dp[idx_2]*2, dp[idx_3]*3, dp[idx_5]*5
		val := min(v1, min(v2, v3))
		dp[count] = val
		count++
		if val == v1 {
			idx_2++
		}
		if val == v2 {
			idx_3++
		}
		if val == v3 {
			idx_5++
		}
	}
	return dp[n-1]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

//----------efficientapproach---------------

func main() {
	fmt.Println(ugly_number_two_eff(10))
}

File: LeetCode/Medium/Ugly Number II/Cpp/ugly_number_two.cc
#include <iostream>

int ugly_number_two(int n) {
    int *dp = new int[n];
    dp[0] = 1;
    int count = 1;
    int ind_2 = 0, ind_3 = 0, ind_5 = 0;
    while(count < n) {
        int v1 = dp[ind_2]*2, v2 = dp[ind_3]*3, v3 = dp[ind_5]*5;
        int val = std::min(v1, std::min(v2, v3));
        dp[count] = val;
        count++;
        if (v1 == val) {
            ind_2++;
        }
        if (v2 == val) {
            ind_3++;
        }
        if (v3 == val) {
            ind_5++;
        }
    }
    return dp[n-1];
}

int main() {
    std::cout<<ugly_number_two(10)<<"\n";
}
    
